test/Microsoft.ML.AutoML.Tests/DatasetUtil.cs

```diff
@@ -49,7 +49,8 @@ public static IDataView GetUciAdultDataView()
 
         private static string DownloadIfNotExists(string baseGitPath, string dataFile)
         {
-            foreach (var nextIteration in Enumerable.Range(0, 10))
```

Comment:
I'd recommend replacing `DownloadIfNotExists()` w/ a call to [`ResourceManagerUtils.Instance.EnsureResourceAsync()`](https://github.com/dotnet/machinelearning/blob/d26b896da20a6de152a68e07a9b27969dbb63fe0/src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs#L87-L114) which at the time it was written, by @yaeldekel, I believed it to have no race conditions when the storage is local, but not for network shares due to lack of a remote atomic rename. It's designed as a thread safe downloader of large files.

Example of call to `EnsureResourceAsync()`:
https://github.com/dotnet/machinelearning/blob/d26b896da20a6de152a68e07a9b27969dbb63fe0/src/Microsoft.ML.TensorFlow/TensorflowUtils.cs#L179-L194

As a background I think this DownloadIfNotExists() function was added

Reply:
Hi @justinormont , indeed there is a race condition here in DownloadIfNotExists(). If ResourceManagerUtils.Instance.EnsureResourceAsync() can indeed download these files in a thread-safe manner, I'll replace `client.DownloadFile` with this. Thank you!

Reply:
Hey Justin, with `ResourceManagerUtils.Instance.EnsureResourceAsync()`, I'm seeing tests that are at some time using this function in the call stack are hanging. I've noticed that some of our TensorFlowImageClassification tests that are currently disabled due to timeout/hanging issues are also using  `ResourceManagerUtils.Instance.EnsureResourceAsync()`. Perhaps these two happenings are related, and that EnsureResourceAsync() is not actually thread-safe?

Reply:
I added an alternative method for checking for bad aka.ms links in your other PR:
https://github.com/dotnet/machinelearning/pull/4943/files#r392572981

I did notice we were retrying bad aka.ms links multiple times, causing a delay of 50 seconds (5 retries x 10 sec). This only affects bad URLs though.

**Testing:**
You could test `ResourceManagerUtils` locally by creating/running ~100 in parallel. This should simulate what's occurring in the CI.

**Singleton:**
`ResourceManagerUtils` is a singleton: 
https://github.com/dotnet/machinelearning/blob/d26b896da20a6de152a68e07a9b27969dbb63fe0/src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs#L23-L32

It's not obvious to me why it's a singleton; I think we could just make all functions static in the class.

**Deadlocks:**
I don't see any deadlocks from the mutex.

**Stall:**
In `return (await t).Message;`, we have potentially timed out via `timeoutTask`, but will still wait for a response from `DownloadResource()`:
https://github.com/dotnet/machinelearning/blob/d26b896da20a6de152a68e07a9b27969dbb63fe0/src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs#L151-L162

We do send the cancellation token to `DownloadResource()`, but it doesn't check the cancellation token until after it receives the next 4096 bytes:
https://github.com/dotnet/machinelearning/blob/d26b896da20a6de152a68e07a9b27969dbb63fe0/src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs#L285-L296

Hence if the server stops sending data, I think this will wait indefinitely. To fix, we could use `Stream.ReadAsync()`, which takes a cancellation token instead of the current `Read()`.

**Other Aync:**
I'm not an expert in C# async code. You may want to look closer at its usage in `ResourceManagerUtils`.